tokio/build.rs

```diff
@@ -0,0 +1,43 @@
+use std::{env, ffi::OsString, process::Command};
+
+fn main() {
+    match rustc_minor_version() {
+        // If rustc >= 1.47.0, we can enable `track_caller`.
+        Ok(minor) if minor >= 47 => println!("cargo:rustc-cfg=tokio_track_caller"),
+        Err(e) => println!("cargo:warning=could not parse rustc version: {}", e),
+        _ => {}
+    }
+}
+
+fn rustc_minor_version() -> Result<usize, Box<dyn std::error::Error>> {
+    let rustc = env::var_os("RUSTC").unwrap_or_else(|| OsString::from("rustc"));
+    let out = Command::new(rustc).arg("-V").output()?;
+    let version_str = std::str::from_utf8(&out.stdout)?;
+    let mut parts = version_str.split(' ');
+    if parts.next() != Some("rustc") {
+        Err(format!(
+            "weird rustc version: {:?} (missing 'rustc') ",
+            version_str
+        ))?;
+    }
+    if let Some(part) = parts.next() {
+        let mut parts = part.split('.');
+        if parts.next() != Some("1") {
+            Err(format!(
+                "weird rustc version: {:?} (does not start with 1)",
+                version_str
+            ))?;
+        }
+        if let Some(middle) = parts.next() {
+            Ok(middle.parse()?)
+        } else {
+            Err(format!("weird rustc version: {:?} (no minor version)", version_str).into())
+        }
+    } else {
+        Err(format!(
+            "weird rustc version: {:?} (missing version entirely?) ",
+            version_str,
+        )
+        .into())
+    }
+}
```

Comment:
alternatively, we could use the [`rustc_version` crate](https://crates.io/crates/rustc_version) for this. I didn't use it, since the parsing we want to do is fairly simple and it didn't seem necessary to add a new build dependency just for parsing the rustc minor version. since it would be a build dep, it would be pulled for all builds, regardless of feature flags, so I thought it was better to do it this way.

if folks disagree, i'm happy to replace this with the `rustc_version` crate!

Reply:
If you want to go the external crate route, I believe that autocfg may be preferable to rustc_version: https://github.com/sfackler/rust-openssl/pull/1103

Reply:
good to know, thanks! i'm fine with adding a dep for this if folks prefer.

Reply:
I would prefer to use autocfg. It's very lightweight. And rust version parsers that parse the output of `rustc --version` instead of `rustc --version --verbose`, including this PR's implementation, may not be able to parse the version of rustc installed by any way other than rustup (e.g., `rustc_version`, `version_check`, etc.)

FWIW, I've actually received [bug reports about this in my project that was using version_check in the past...](https://github.com/taiki-e/const_fn/issues/24#issuecomment-684180738) (I'm using [own parser](https://github.com/taiki-e/const_fn/blob/ddc39258c3b3084c9d4ae721f0b1181824d9c2a2/build.rs#L36-L39) for now because [autocfg doesn't provide the ability to detect the nightly compiler](https://github.com/cuviper/autocfg/issues/28), but I think it makes sense to use autocfg if it doesn't need that ability.)

Reply:
> AFAICT, it isn't possible to feature flag build deps, so we will always pull `autocfg`, even when the tracing feature is disabled...i'm not sure if that's a huge issue.

(as you said in another comment, this PR is currently always using `track_caller`)